core/js/src/UtcOffset.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2019-2021 JetBrains s.r.o.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.datetime
+
+import kotlinx.datetime.internal.JSJoda.ZoneOffset
+import kotlinx.datetime.serializers.UtcOffsetSerializer
+import kotlinx.serialization.Serializable
+
+@Serializable(with = UtcOffsetSerializer::class)
+public actual class UtcOffset(internal val zoneOffset: ZoneOffset) {
+    public actual val totalSeconds: Int get() = zoneOffset.totalSeconds().toInt()
+
+    override fun hashCode(): Int = zoneOffset.hashCode().toInt()
+    override fun equals(other: Any?): Boolean = other is UtcOffset && this.zoneOffset == other.zoneOffset
+    override fun toString(): String = zoneOffset.toString()
+
+    public actual companion object {
+
+        public actual val ZERO: UtcOffset = UtcOffset(ZoneOffset.UTC)
+
+        public actual fun parse(offsetString: String): UtcOffset = try {
+            ZoneOffset.of(offsetString).let(::UtcOffset)
+        } catch (e: Throwable) {
+            if (e.isJodaDateTimeException()) throw DateTimeFormatException(e)
+            throw e
+        }
+    }
+}
+
+@Suppress("ACTUAL_FUNCTION_WITH_DEFAULT_ARGUMENTS")
+public actual fun UtcOffset(hours: Int = 0, minutes: Int = 0, seconds: Int = 0): UtcOffset =
+    try {
+        if (hours == 0 && minutes == 0) {
+            UtcOffset(ZoneOffset.ofTotalSeconds(seconds))
+        } else {
+            UtcOffset(ZoneOffset.ofHoursMinutesSeconds(hours, minutes, seconds))
+        }
+    } catch (e: Throwable) {
+        if (e.isJodaDateTimeException()) throw IllegalArgumentException(e) else throw e
+    }
```

Comment:
It's interesting to note that java.time doesn't have a function like `ofTotalMinutes()`, yet there are no questions about how to construct offset from minutes. Maybe it's an evidence that it's not actually needed, or that it's obvious that `ofTotalSeconds(minutes * 60)` can be used instead.

Reply:
Sure, but the difference is that Java's approach is consistent: they provide a group of functions that behave in an expected way, though, yes, some functions that could have been implemented are missing. Discontinuity, in my opinion, is bad regardless of whether it can be worked around and whether it is triggered commonly. I could elaborate with specific examples of potential problems this discontinuity could cause if you'd like me to.

Reply:
I feel that this constructor function contract is already overloaded with two ways of checking parameter ranges, and not sure that it's worth adding another way to the mix.

Please share your concerns about discontinuity.

Reply:
This contract could be simplified to just "everything that comes after the first non-zero component gets range-checked". Seems succinct. Or we could prohibit the `Utc(seconds = 7200)` usage for now and possibly add it in some form if there's demand.

As for concerns, consider this hypothetical. Let's say a user attempts to use this function with 180 minutes. The system complains, talking about range-checks. The user infers that, surely, this means that this function behaves in the same way as `ofHoursMinutesSeconds` in Java, that is, it expects the arguments to reflect the human-readable representation of the offset. Then, encountering the form `UtcOffset(seconds = 7200)`, they believe that it's a bug. And much, much smaller things have led to users losing their sense of flow and becoming greatly confused.

tl;dr: other APIs for `UtcOffset` equivalents and for similar things (like, in a sense, `LocalTime`) provide constructors that range-check, and the user is at risk of classifying this constructor as such if they happen to encounter range-checking of minutes.

Reply:
I agree that the proposed change would simplify the contract of this constructor, but this comes at the expense of not enforcing the same value ranges as `ofHoursMinutesSeconds` in Java or the `UtcOffset.parse` function do.
So for example, if a user has parsed the components of an offset like `00:60:00` manually and now wants to construct an offset validating the range of these components, `UtcOffset` function would be of no help here.

The problem lies in the fact that the missing hours component is indistinguishable from the present hours that is equal to zero. This brings an idea that we could use nullable Int parameters and `null` as a default value to distinguish these two situations.

Reply:
> So for example, if a user has parsed the components of an offset like `00:60:00` manually and now wants to construct an offset validating the range of these components, `UtcOffset` function would be of no help here.

The same can be said about `00:00:60`. That's the problem with the discontinuity: minutes and seconds are not inherently different, and if we behave like they are, it's inconsistent and confusing.

> This brings an idea that we could use nullable Int parameters and `null` as a default value to distinguish these two situations.

Could you clarify the behavior that you invision here? I don't understand what `null` would mean. Also, which parameters would be nullable?

Reply:
Basically it means the same thing as you proposed "everything that comes after the first non-zero component gets range-checked", except of "non-null" instead of "non-zero".

Reply:
This looks semantically clean. I don't know if such a solution is idiomatic in Kotlin, but you're a better judge of that than I am.